omnipaxos_core/src/storage.rs

```diff
@@ -468,62 +475,52 @@ where
     // accepted index (not including the batched entries)
     pub(crate) fn append_entry(&mut self, entry: T) -> Option<(u64, Vec<T>)> {
         let append_res = self.state_cache.append_entry(entry);
-        if let Some(flushed_entries) = append_res {
-            self.storage.append_entries(flushed_entries.clone());
-            Some((self.get_accepted_idx(), flushed_entries))
-        } else {
-            None
-        }
+        append_res.map(|flushed_entries| (self.append_entries_without_batching(flushed_entries.clone()), flushed_entries))
     }
 
     // Append entries in batch, if the batch size is reached, flush the batch and return the
     // accepted index. If the batch size is not reached, return None.
     pub(crate) fn append_entries_with_batching(&mut self, entries: Vec<T>) -> Option<u64> {
         let append_res = self.state_cache.append_entries(entries);
-        if let Some(flushed_entries) = append_res {
-            self.storage.append_entries(flushed_entries);
-            Some(self.get_accepted_idx())
-        } else {
-            None
-        }
+        append_res.map(|flushed_entries| self.append_entries_without_batching(flushed_entries))
     }
 
     // Append entries without batching, return the accepted index
     pub(crate) fn append_entries_without_batching(&mut self, entries: Vec<T>) -> u64 {
         self.storage.append_entries(entries);
+        self.update_accepted_idx();
         self.get_accepted_idx()
```

Comment:
The `updated_accepted_idx()` function reads from the persistent storage to get the log length. This is slow and unnecessary since we already get back the real_log length from `self.storage.append_entries(entries)`:

```suggestion
        let real_log_len = self.storage.append_entries(entries);
        let accepted_idx = real_log_len + self.state_cache.get_compacted_idx();
        self.state_cache.accepted_idx = accepted_idx;
        accepted_idx
```